appinventor/blocklyeditor/src/generators/yail/control.js

```diff
@@ -93,6 +93,50 @@ Blockly.Yail['controls_forEach'] = function() {
          + listCode + Blockly.Yail.YAIL_CLOSE_COMBINATION;
 };
 
+Blockly.Yail['controls_for_each_dict'] = function() {
+  var yail = Blockly.Yail;
+  var prefix = Blockly.usePrefixInYail ? 'local_' : '';
+  var keyName = yail.YAIL_LOCAL_VAR_TAG + prefix + this.getFieldValue('KEY');
+  var valueName = yail.YAIL_LOCAL_VAR_TAG + prefix + this.getFieldValue('VALUE');
+
+  var dictionaryCode = yail.valueToCode(this, 'DICT', yail.ORDER_NONE);
+  var bodyCode = yail.statementToCode(this, 'DO') || yail.YAIL_FALSE;
+
+  var loopIndexName = 'item';
+  var loopIndexCommandAndName = yail.getVariableCommandAndName(loopIndexName);
+  loopIndexName = loopIndexCommandAndName[1];
+  var getListCode = loopIndexCommandAndName[0] + loopIndexName +
+      yail.YAIL_CLOSE_COMBINATION;
+  var getKeyCode = yail.YAIL_CALL_YAIL_PRIMITIVE + 'yail-list-get-item' + yail.YAIL_SPACER
+      + yail.YAIL_OPEN_COMBINATION + yail.YAIL_LIST_CONSTRUCTOR + yail.YAIL_SPACER
+      + getListCode
+      + yail.YAIL_SPACER + '1' + yail.YAIL_CLOSE_COMBINATION
+      + yail.YAIL_SPACER + yail.YAIL_QUOTE + yail.YAIL_OPEN_COMBINATION
+      + 'list number' + yail.YAIL_CLOSE_COMBINATION + yail.YAIL_SPACER
+      + yail.YAIL_DOUBLE_QUOTE + 'select list item' + yail.YAIL_DOUBLE_QUOTE + yail.YAIL_CLOSE_COMBINATION;
+  var setKeyCode = yail.YAIL_LET + yail.YAIL_OPEN_COMBINATION + yail.YAIL_SPACER
+      + yail.YAIL_OPEN_COMBINATION + keyName
+      + yail.YAIL_SPACER + getKeyCode
+      + yail.YAIL_CLOSE_COMBINATION + yail.YAIL_SPACER;
+  var getValueCode = yail.YAIL_CALL_YAIL_PRIMITIVE + 'yail-list-get-item' + yail.YAIL_SPACER
+      + yail.YAIL_OPEN_COMBINATION + yail.YAIL_LIST_CONSTRUCTOR + yail.YAIL_SPACER
+      + getListCode
+      + yail.YAIL_SPACER + '2' + yail.YAIL_CLOSE_COMBINATION
+      + yail.YAIL_SPACER + yail.YAIL_QUOTE + yail.YAIL_OPEN_COMBINATION
+      + 'list number' + yail.YAIL_CLOSE_COMBINATION + yail.YAIL_SPACER
+      + yail.YAIL_DOUBLE_QUOTE + 'select list item' + yail.YAIL_DOUBLE_QUOTE + yail.YAIL_CLOSE_COMBINATION;
+  var setValueCode = yail.YAIL_SPACER
```

Comment:
Ok. So on further inspection of your log I think I've found the potential problem. In the `setKeyCode` string you open the `(let (` but you never close the variable definition. You'll want to have two `YAIL_CLOSE_COMBINATION` here, one to close the `($value ...)` and a second to close the `(let (` started in `setKeyCode`. The generic form of the let is:

```scheme
(let ((var1 value1) (var2 value2) ...) code ...)
```

but since you don't have the closing parens after the variable definitions you end up with unbalanced parens, resulting in the EOF error.

Reply:
Awesome! Thank you for the help :D I'll take another shot at it later today!